Adding ssht package

mdavezac opened the pull request:
Fixes #4819

Specify library name and version:  **ssht/1.3.6**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=4820) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/conan-center-index?pullRequest=4820) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/conan-center-index?pullRequest=4820) it.</sub>

prince-chrismc reviewed (changes requested):
Thanks for opening this PR, we appreciate the contribution. 

Since this is you first PR, I laid out lots of comments but there's also the export sources and build methods which does not support all the workflows we needs to support

I'd like to encourage you to look at https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md#how-to-provide-a-good-recipe

Personal I think this is a great example close to this project https://github.com/conan-io/conan-center-index/blob/master/recipes/cryptopp/all/conanfile.py

prince-chrismc commented on recipes/ssht/all/test_package/conanfile.py:
```diff
@@ -15,6 +15,6 @@ def build(self):
-            os.chdir("bin")
-            self.run(".%sexample" % os.sep)
+        if not tools.cross_building(self.settings):
+            bin_path = os.path.join("bin", "test_package")
```
Sorry for my bad copy pasting

```suggestion
            bin_path = os.path.join("bin", "example")
```

prince-chrismc commented on recipes/ssht/all/conanfile.py:
```diff
@@ -14,53 +14,51 @@ class SshtConan(ConanFile):
 
-    def source(self):
-        tools.get(**self.conan_data["sources"][self.version])
+        self.cpp_info.names["cmake_find_package"] = "SSHT"
+        self.cpp_info.names["cmake_find_package_multi"] = "SSHT"
```
There's no namespace upstream https://github.com/astro-informatics/ssht/blob/061763ac1140fb77b16301ef45fe1069b1f2795c/src/c/CMakeLists.txt#L19
This is not required to be implemented https://github.com/conan-io/conan-center-index/blob/770d232151108426f4fadf48e2967434865c96a0/recipes/tesseract/all/conanfile.py#L131 but please add a FIXME so we know


This does not match the `FindPackage(<NAME>)` https://github.com/astro-informatics/ssht/blob/061763ac1140fb77b16301ef45fe1069b1f2795c/cmake/exporting.cmake#L12
the `cpp_info` has a `filenames` for that (same syntax as names)

mdavezac commented on recipes/ssht/all/conandata.yml:
```diff
@@ -1,4 +1,4 @@
-  "1.3.6":
-    url: "https://github.com/astro-informatics/ssht/archive/1.3.6.zip"
-    sha256: "ce6b822e755ed1ea224cce0a9e557b035506c27d3828427ef007009fdd0c5357"
+  "1.3.7":
+    url: "https://github.com/astro-informatics/ssht/archive/1.3.7rc1.zip"
```
Yes, before we merge this, I'd like to do a full release.

prince-chrismc replied:
just a reminder it's still a pre-release

mdavezac commented on recipes/ssht/all/conanfile.py:
```diff
@@ -0,0 +1,68 @@
+    options = {"fPIC": [True, False]}
+    default_options = {"fPIC": True}
+    requires = "fftw/3.3.9"
+    generators = "cmake"
```
I don't think we need cmake_find_package. The  line you pointed to is not exercised if conan is used for handling dependencies.

mdavezac commented on recipes/ssht/all/conanfile.py:
```diff
@@ -0,0 +1,68 @@
+    description = "Fast spin spherical harmonic transforms"
+    settings = "os", "arch", "compiler", "build_type"
+    topics = ("Physics", "Astrophysics", "Radio Interferometry")
+    options = {"fPIC": [True, False]}
+    default_options = {"fPIC": True}
```
Ah... yeah... I would rather keep that can of worms firmly closed for now, if that's okay.

prince-chrismc replied:
That's acceptable!

mdavezac commented:
I've added a namespace upstream and moved to using cmake_find_packages and cmake_paths since it's a bit less disruptive.

I think I've dealt with all your comments. Thank you so much for the hand-holding. It's very much appreciated.
If everything is OK, I'll do a release upstream and change the conandata.yml file appropriately.

prince-chrismc commented:
Please do not force push it makes it more difficult to review (since we need to restart)

prince-chrismc commented on recipes/ssht/all/test_package/conanfile.py:
```diff
@@ -0,0 +1,20 @@
+
+class SshtTestConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake", "cmake_find_package", "cmake_paths"
```
Typically we use https://github.com/conan-io/conan-center-index/blob/eb5f141d056d92020e3bee7bb4f9eeca52633a3f/recipes/any-lite/all/test_package/conanfile.py#L7

```suggestion
    generators = "cmake", "cmake_find_package"
```

prince-chrismc commented on recipes/ssht/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.12)
+project(PackageTest C)
+
+include(${CMAKE_BINARY_DIR}/conan_paths.cmake)
+find_package(ssht)
+add_executable(test_package example.c)
+target_link_libraries(test_package ssht::ssht)
+target_compile_features(test_package PUBLIC c_std_99)
+set_target_properties(test_package PROPERTIES RUNTIME_OUTPUT_DIRECTORY
+                                              "${PROJECT_BINARY_DIR}/bin")
```
then the rest is handled by our template... `conan_basic_setup()` does the magic

```suggestion
cmake_minimum_required(VERSION 3.1)
project(test_package CXX)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup()

find_package(Ssht REQUIRED)

add_executable(${PROJECT_NAME} example.c)
target_compile_features(${PROJECT_NAME} PUBLIC c_std_99)
target_link_libraries(${PROJECT_NAME} ssht::ssht)
```

prince-chrismc commented on recipes/ssht/all/conanfile.py:
```diff
@@ -0,0 +1,59 @@
+        self.copy("LICENSE", dst="licenses", src=self._source_subfolder)
+        self.cmake.install()
+
+    def package_info(self):
```
You'll need to set the `filenames` for https://github.com/astro-informatics/ssht/blob/main/cmake/SshtConfig.in.cmake to be generated correctly by conan

mdavezac replied:
I'm not sure I understand this comment. As far as I can tell, conan generates both xxxConfig.cmake and xxxTargets.cmake in the build directory. I've actually removed generating those upstream, since conan did not pick them up.

prince-chrismc commented on recipes/ssht/all/conanfile.py:
```diff
@@ -0,0 +1,59 @@
+
+    def package_info(self):
+        self.cpp_info.libs = ["ssht"]
+        self.cpp_info.builddirs = [os.path.join("lib", "cmake", "ssht")]
```
Ahhh!!

prince-chrismc reviewed (changes requested):
There's one issue.. it's related to this in our docs

https://github.com/conan-io/conan-center-index/blob/master/docs/faqs.md#why-are-cmake-findconfig-files-and-pkg-config-files-not-packaged

prince-chrismc commented on recipes/ssht/all/test_package/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package C)
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+find_package(Ssht REQUIRED NO_MODULE)
+add_executable(${PROJECT_NAME} example.c)
+target_compile_features(${PROJECT_NAME} PUBLIC c_std_99)
+target_link_libraries(${PROJECT_NAME} ssht::ssht)
+cmake_minimum_required(VERSION 3.12)
+project(PackageTest C)
```
```suggestion
cmake_minimum_required(VERSION 3.1)
project(test_package C)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup()

find_package(Ssht REQUIRED NO_MODULE)

add_executable(${PROJECT_NAME} example.c)
target_compile_features(${PROJECT_NAME} PUBLIC c_std_99)
target_link_libraries(${PROJECT_NAME} ssht::ssht)
```

prince-chrismc commented:
Please do not force push it makes it more difficult to review

mdavezac commented:
@prince-chrismc, thanks again for all the help! Is it right that we just need the upstream to go from the current rc to a full release at this point?

prince-chrismc commented:
Yes!

prince-chrismc commented:
Please do not force push 🙏

prince-chrismc commented:
Shouldn't be long before other review and it gets merged.

If theres nothing, in 2 days add a comment I'll pick the other reviewers (this has dropped in the list might be mised)

prince-chrismc commented:
Oh remove draft! That willhelp

mdavezac commented:
Thanks again for all the help @prince-chrismc. I really appreciate it!

prince-chrismc commented:
Always a pleasure :hugs:
